torch/csrc/serialization.cpp

```diff
@@ -8,3 +8,129 @@
+inline ssize_t doRead<int>(int fildes, void* buf, size_t nbytes) {
+  return read(fildes, buf, nbytes);
+}
+
+template <>
+inline ssize_t doRead<PyObject*>(PyObject* fildes, void* buf, size_t nbytes) {
+  // Try to use fildes.readinto() instead of fildes.read()
+  // because it is more memory efficient.
+  auto has_readinto = PyObject_HasAttrString(fildes, "readinto") == 1;
+  if (has_readinto) {
+    return doPythonReadInto(fildes, buf, nbytes);
+  }
+  return doPythonReadBuffered(fildes, buf, nbytes);
+}
+
+template <>
+inline ssize_t doWrite<int>(int fildes, void* buf, size_t nbytes) {
+  return write(fildes, buf, nbytes);
+}
+
+template <>
+inline ssize_t doWrite<PyObject*>(PyObject* fildes, void* buf, size_t nbytes) {
+  return doPythonWrite(fildes, buf, nbytes);
+}
+
+static inline bool isUnsupportedOperation() {
+  THPObjectPtr io(PyImport_ImportModule("io"));
+  if (!io) throw python_error();
+  THPObjectPtr exception(PyObject_GetAttrString(io, "UnsupportedOperation"));
+  if (!exception) python_error();
+  return PyErr_ExceptionMatches(exception.get());
+}
+
+// Call Python fildes.read(nbytes) and copy it to buf.
+inline ssize_t doPythonReadBuffered(PyObject* fildes, void* buf, size_t nbytes) {
+  const size_t buffer_size = 262144;  // 2^18
+  size_t read_bytes = 0;
+
+  while (read_bytes < nbytes) {
+    auto remaining = nbytes - read_bytes;
+    auto to_read = remaining > buffer_size ? buffer_size : remaining;
+    THPObjectPtr r(PyObject_CallMethod(fildes, "read", "i", to_read));
+    if (!r) throw python_error();
+
+    // read output is String (Python 2) / Bytes (Python 3)
+#if PY_MAJOR_VERSION >= 3
+    auto size = PyBytes_GET_SIZE(r.get());
+    const void* bytes = PyBytes_AsString(r.get());
+#else
+    auto size = PyString_GET_SIZE(r.get());
+    const void* bytes = PyString_AsString(r.get());
+#endif
+
+    // we read EOF
+    if (size == 0) {
+      return read_bytes;
+    }
+
+    memcpy(reinterpret_cast<char*>(buf) + read_bytes, bytes, size);
+    read_bytes += size;
+  } // Reading loop
+
+  return read_bytes;
+}
+
+// Either does fildes.readinto(buf) or fildes.write(buf)
+static inline ssize_t doPythonIO(PyObject* fildes, void* buf, size_t nbytes, bool is_read) {
+#if PY_MAJOR_VERSION >= 3
+  auto rw_flag = is_read ? PyBUF_WRITE : PyBUF_READ;
+  THPObjectPtr memview(PyMemoryView_FromMemory(
+      reinterpret_cast<char*>(buf), nbytes, rw_flag));
+#else
+  // PyMemoryView_FromMemory doesn't exist in Python 2.7, so we manually
+  // create a Py_buffer that describes the memory and create a memoryview from it.
+  auto readonly_flag = is_read ? 1 : 0;
+  Py_buffer pyBuf;
+  pyBuf.buf = buf;
+  pyBuf.obj = nullptr;
+  pyBuf.len = (Py_ssize_t)nbytes;
+  pyBuf.itemsize = 1;
+  pyBuf.readonly = readonly_flag;
+  pyBuf.ndim = 0;
+  pyBuf.format = nullptr;
+  pyBuf.shape = nullptr;
+  pyBuf.strides = nullptr;
+  pyBuf.suboffsets = nullptr;
+  pyBuf.internal = nullptr;
+
+  THPObjectPtr memview(PyMemoryView_FromBuffer(&pyBuf));
```

Comment:
This part of the docs for `PyMemoryView_FromBuffer` is a bit worrying for me:
> The memoryview object then owns the buffer, which means you shouldn’t try to release it yourself: it will be released on deallocation of the memoryview object.

Why not use `PyBuffer_FromReadWriteMemory`?

Reply:
Good point, I didn't catch `PyBuffer_FromReadWriteMemory` when I was going through the docs. Let me try to use it.